website/sales/api/v2/serializers/order.py

```diff
@@ -0,0 +1,168 @@
+
+from members.api.v2.serializers.member import MemberSerializer
+from payments.api.v1.serializers import PaymentSerializer
+from sales.models.order import Order, OrderItem
+from sales.models.product import ProductListItem
+
+
+class ProductNameRelatedField(serializers.SlugRelatedField):
+    def get_queryset(self):
+        shift = self.root.context.get("shift", None)
+        if shift is None:
+            shift = self.root.instance.shift
+        return ProductListItem.objects.filter(product_list=shift.product_list)
+
+    def to_internal_value(self, data):
+        if type(data) is ProductListItem:
+            return data
+
+        queryset = self.get_queryset()
+        try:
+            return queryset.get(product__name=data)
+        except ObjectDoesNotExist:
+            self.fail(
+                "does_not_exist", slug_name=self.slug_field, value=smart_str(data)
+            )
+        except (TypeError, ValueError):
+            self.fail("invalid")
+
+    def to_representation(self, obj):
+        return obj.product.name
+
+
+class OrderItemSerializer(serializers.ModelSerializer):
+    """Serializer for order items."""
+
+    class Meta:
+        model = OrderItem
+        fields = ("product", "amount", "total")
+        read_only_fields = (
+            "total",
+        )  # TODO make this writable if user has custom_prices permission
+
+    product = ProductNameRelatedField("product")
+
+    def create(self, validated_data, **kwargs):
+        order = self.context["order"]
+        item = OrderItem.objects.create(order=order, **validated_data)
+        return item
+
+    def update(self, instance, validated_data, **kwargs):
+        order = self.context["order"]
+        instance.order = order
+        instance.total = None  # Always recalculate the total amount if updating using API (note the difference from the model that only recalculates if the total is None, to deal with historic data and allow for special discounts)
+        super().update(instance, validated_data)
+        return instance
+
+
+class OrderSerializer(serializers.ModelSerializer):
+    """Serializer for orders."""
+
+    class Meta:
+        model = Order
+        fields = (
+            "pk",
+            "shift",
+            "created_at",
+            "order_items",
+            "order_description",
+            "age_restricted",
+            "subtotal",
+            "discount",
+            "total_amount",
+            "num_items",
+            "payment",
+            "payer",
+            "payment_url",
+        )
+        read_only_fields = (
+            "pk",
+            "shift",
+            "created_at",
+            "discount",  # TODO make this writable if user has custom_prices permission
+            "payment",
+            "total_amount",
+            "num_items",
+            "age_restricted",
+            "order_description",
+            "payment_url",
+        )
+
+    shift = serializers.PrimaryKeyRelatedField(read_only=True)
```

Comment:
Isn't is possible to make this the ShiftSerializer directly?

Reply:
I figured that would just be too much data to send that's not really required to know